src/opuntia.erl

```diff
@@ -132,27 +140,81 @@ calculate(#token_bucket_shaper{shape = {MaximumTokens, Rate},
     TokensOverused = max(0, TokensNowUsed - ExactlyAvailableNow),
 
     %% And then MaybeDelay will be zero if TokensOverused was zero
-    MaybeDelayMs = TokensOverused / Rate,
+    OverUsedRateNow = TokensOverused / Rate,
+
+    NewShaper = Shaper#token_bucket_shaper{available_tokens = TokensAvailable},
+    Punish = OverUsedRateNow - OverPenalisedInUnitsLastTime,
+    final_state(NewShaper, TimeUnit, Punish, NativeNow).
+
+-spec timediff_in_units(time_unit(), integer(), integer()) -> float().
+timediff_in_units(TimeUnit, NativeLastUpdate, NativeNow) ->
+    %% Time difference between now and the last update, in native
+    NativeTimeSinceLastUpdate = NativeNow - NativeLastUpdate,
+    %% This is now a float and so will all below be, to preserve best rounding errors possible
+    convert_time_unit(NativeTimeSinceLastUpdate, native, TimeUnit).
 
-    %% We penalise rounding up, the most important contract is that rate will never exceed that
-    %% requested, but the same way timeouts in Erlang promise not to arrive any time earlier but
-    %% don't promise at what time in the future they would arrive, nor we promise any upper bound
-    %% to the limits of the shaper delay.
-    RoundedDelayMs = ceil(MaybeDelayMs),
+%% Unbounded growth is calculated, with float precision, in the configured time units
+%%
+%% Note that it can be negative, if earlier we have penalised giving a larger `last_update' and now we
+%% update even before we have reach the point in time where the previous `last_update' was set
+%%
+%% If the growth was negative, that means that it has grown the debt instead
+-spec unbounded_bucket_diff(rate(), tokens(), float()) -> float().
+unbounded_bucket_diff(Rate, LastAvailableTokens, TimeDiffInUnits) ->
```

Comment:
Just an opinion about naming.

Why is this called `diff`? To me diff is `AvailableAtGrowthRate`, and `LastAvailableTokens + AvailableAtGrowthRate` is the total amount of tokens rather than a diff.

It is just my point of view, but I would make the function naming more consistent, like `unbounded_available_tokens` and `exactly_available_tokens`.